### ISSUE TITLE
HttpListener - why queue polling ?

### ISSUE THREAD

**Issue by MrDigit** [unosquare/embedio#155]
Hello,

first, thanks for the great project. Just looked through the code in HttpListener.cs and stumbled over this code:

        public async Task<HttpListenerContext> GetContextAsync()
        {
            while (true)
            {
                foreach (var key in _ctxQueue.Keys)
                {
                    if (_ctxQueue.TryRemove(key, out var context))
                        return context;
                }

                await Task.Delay(10);
            }
        }

From my understanding polling every 10ms is not really a good idea: **await Task.Delay(10)**.
Why not doing something like this:

        private System.Threading.SemaphoreSlim _ctxQeueSem = new System.Threading.SemaphoreSlim(0);

        public async Task<HttpListenerContext> GetContextAsync()
        {
            while (true)
            {
                await _ctxQeueSem.WaitAsync();

                foreach (var key in _ctxQueue.Keys)
                {
                    if (_ctxQueue.TryRemove(key, out var context))
                        return context;
                    break;
                }
            }
        }

        internal void RegisterContext(HttpListenerContext context)
        {
            if (_ctxQueue.TryAdd(context.Id, context) == false)
                throw new Exception("Unable to register context");

            _ctxQeueSem.Release();
        }

Good idea, bad idea ? Any feedback appreciated :)

Best regards
  MrDigit

**Comment by geoperez** [comment]
Let me take a look. Thank you.

**Comment by geoperez** [comment]
Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.8      0       8
Processing:     1    9  15.6      5     208
Waiting:        0    8  15.4      4     205
Total:          1   10  15.7      6     210

**Comment by geoperez** [comment]
After the change

step|min| mean|[+/-sd]|median|max
-- | -- | -- | -- | -- | --
Connect:   |     0|    3|   4.3 |     2 |     40
Processing:  |   1 |  12 | 10.3  |    9  |    78
Waiting:      |  0  | 10|   9.8  |    7   |   75
Total:    |      1  | 15  |12.3 |    12  |    89

step|min| mean|[+/-sd]|median|max
-- | -- | -- | -- | -- | --
Connect:     |   0 |   1 |  0.8   |   0 |      6
Processing:   |     1   |    8  |   16.6   |      4   |     259
Waiting:      |    0   |    8  |   16.5    |     3  |      259
Total:    |         1   |    9  |   16.7    |     4     |   262

step|min| mean|[+/-sd]|median|max
-- | -- | -- | -- | -- | --
Connect:    |    0    |1  | 0.9   |   0   |    6
Processing:|     1 |   8  |19.5     | 4 |    285
Waiting:   |     0|    7|  19.3    |  3 |    285
Total:      |    1   | 9 | 19.5   |   5  |   286


step|min| mean|[+/-sd]|median|max
-- | -- | -- | -- | -- | --
Connect:    |    0  |  1  | 1.9 |     0 |     23
Processing:   |  1  |  8|  18.3 |     4  |   353
Waiting:     |   0|    7 | 18.1    |  3   |  352
Total:     |     1   | 9|  18.4   |   5  |   354

**Comment by jpalcala** [comment]
using : `ab -n 100000 -c 15 http://localhost:8787/`


p| Previous method | New method
------------ | --------------|--------------
Time taken for tests | 111.180 seconds |118.108 seconds
Requests per second | 899.44 [#/sec] (mean) |  846.69 [#/sec] (mean)
Time per request |16.677 [ms] (mean) |17.716 [ms] (mean)
Transfer rate | 2094.89 [Kbytes/sec] received|1972.02 [Kbytes/sec] received


Before changes:

step | min | mean | [+/-sd] | median | max
-----|-----|-------|---------|--------|-----
Connect:|     0  |  1  | 1.7  |    0   |   45
Processing:   |  1  | 15  |41.6   |   6   | 1829
Waiting:     |   0  | 14 | 41.4   |   5  |  1828
Total:        |  1  | 16  |41.7  |    7   | 1830

After changes:

step | min | mean | [+/-sd] | median | max
-----|-----|-------|---------|--------|-----
Connect:|     0  |  2  | 2.5  |    1  |   99
Processing:   |  1  | 15  |26.9  |   9   | 1021
Waiting:     |   0  | 14 | 26.6   |   7  |  1018
Total:        |  1  | 17 |27.2  |   11   | 1021


tldr: I can't see any conclusive evidence that can support the new approach that you suggested, however I think we'll keep this issue open in case that anyone can think of a better way to achieve this and also if someone has a better way of assessing .

**Comment by MrDigit** [comment]
Thanks for the comparison.

The difference in performance might be due to the fact that with the proposed SemaphoreSlim-Version every request goes through one more lock-statement (as SemaphoreSlim uses locking internally).

In my opinion the version with Task.Delay(10) is bad style as it consumes CPU resources without needs.

If a small performance hit in a synthetic benchmark is a no go for you and your goal is trying to boost performance then my solution is not the right way. Possible way (not tested) would be then:

1) Exchange _ctxQueue class ConcurrentDictionary with a lock-free queue-Version. 
2) Use something like an async AutoReset-Event instead of SemaphoreSlim and wait for that only when the queue is empty not at every Dequeue-request.

If this proposal sounds good to you I can post an example.

**Comment by geoperez** [comment]
Sure, we really appreciate your interest. Please post your code.
